### DISCUSSION TITLE
Running Directus through a Nginx proxy with sub folder strategy?

### DISCUSSION THREAD

**Discussion by Repox** [directus/directus#4480] [Archive]
I can't seem to find any relevant information other than what [the documentation](https://docs.directus.io/guides/installation/ubuntu/) is telling me. 

The documentation points me to a [DO Community article](https://www.digitalocean.com/community/tutorials/how-to-set-up-a-node-js-application-for-production-on-ubuntu-20-04) which explains that to proxy with a sub folder strategy, I can write something like the following:

```nginx
    location /directus {
            proxy_pass http://localhost:3001;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection 'upgrade';
            proxy_set_header Host $host;
            proxy_cache_bypass $http_upgrade;
        }
```

In my `.env` file I've added 

```ini
PUBLIC_URL="https://example.com/directus/"
```

I've also tried with just `/directus`  and `/` just for testing it out.
But the Directus server just responds with the following:

```
13:04:47 ✨ Server started at port 3001
13:04:47 ✨ Email connection established
13:05:16 ✨ request completed GET 404 /directus 36ms
13:05:35 ✨ request completed GET 404 /directus 22ms
13:05:36 ✨ request completed GET 404 /directus 19ms
13:05:36 ✨ request completed GET 404 /directus 18ms
13:05:36 ✨ request completed GET 404 /directus 25ms
13:05:36 ✨ request completed GET 404 /directus 14ms
13:05:39 ✨ request completed GET 404 /directus/admin 23ms
13:05:55 ✨ request completed GET 404 /directus/admin 14ms
```

So what is the correct approach (if possible at all) to have this working with a reverse proxy and a subfolder strategy?

**Comment by Repox** [comment]
So, I made it work, but it feels wonky and I'd wish that it was more of a configurable setting in Directus instead of a quirky fix necessary in the Nginx configuration, which makes the workings of proxying to Directus less transparent - especially considering the external references (mentioned in my original post) that is directly referenced by Directus official documentation.

To make it work, as of the time of writing, you will have to rewrite the request based on your subfolder strategy.

By adding a `rewrite` rule, you can proxy the request to the root path of the running Directus instance, ie.:

```nginx
rewrite ^/{subfolder-name}/(.*)$ /$1 break;
```

 \- where `{subfolder-name}` should be replaced with your desired subfolder name.

In my case, a complete Nginx `location` block would look like this:

```nginx
location /directus {
    rewrite ^/directus/(.*)$ /$1 break;
    proxy_pass http://localhost:3001;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection 'upgrade';
    proxy_set_header Host $host;
    proxy_cache_bypass $http_upgrade;
}
```

The `location` block is unaffected by the `PUBLIC_URL` configuration as this basically just refers to URL building in mails and other route building functions.

I hope the answer above helps others who are trying to achieve similar setup.

**Reply by Nisthar** [reply]
@Repox Why not use host:8055/ to access directus?

**Reply by Repox** [reply]
> @Repox Why not use host:8055/ to access directus?

Oh, gosh. Completely forgot about this issue since I gave up on Directus. Closing as the question is irrelevant.

**Reply by Nisthar** [reply]
> > @Repox Why not use host:8055/ to access directus?
> 
> Oh, gosh. Completely forgot about this issue since I gave up on Directus. Closing as the question is irrelevant.

Why did you gave up on directus? also what solution are you using now?

**Reply by Repox** [reply]
> Why did you gave up on directus? also what solution are you using now?

At the time, it seemed like a solution that was untimely maintained, community support was sparse and in general, the experience was just not what I hoped it to be.

I never found an alternative, so I continued with my needs with Laravel Nova instead.

**Reply by gokaybiz** [reply]
> @Repox Why not use host:8055/ to access directus?

Let's assume there is only one domain, and multiple projects.
Like vue, react and also directus.
When you set directus as subfolder (/api), it's proccessing incoming requests as 127.0.0.1:8055/api

So, that rule stands for prevent from this:
example.com/api -> 127.0.0.1:8055/api
It should be:
example.com/api/$1 -> 127.0.0.1:8055/$1

Thank you so much for your answer @Repox

**Comment by avionbg** [comment]
There is another issue when mounting the app on subfolder trough nginx - client SDK fails to login.
Culprit is in the `node_modules/@directus/sdk/dist/sdk.bundler.js` line ~ 821

```js
this._transport.post('/auth/login'
```

This needs to be manually updated, adding the folder to the path (in previous example it would be post('/directus/auth/login'.
Just as a note for anyone else who gets this issue (I lost several hours debugging it).

**Reply by licitdev** [reply]
@avionbg Mind filing an issue in https://github.com/directus/sdk ?
Thank you!

**Reply by avionbg** [reply]
Done : [Bug report](https://github.com/directus/sdk/issues/51)

**Comment by abhiche** [comment]
[> So, I made it work, but it feels wonky and I'd wish that it was more of a configurable setting in Directus instead of a quirky fix necessary in the Nginx configuration, which makes the workings of proxying to Directus less transparent - especially considering the external references (mentioned in my original post) that is directly referenced by Directus official documentation.
> 
> To make it work, as of the time of writing, you will have to rewrite the request based on your subfolder strategy.
> 
> By adding a `rewrite` rule, you can proxy the request to the root path of the running Directus instance, ie.:
> 
> ```nginx
> rewrite ^/{subfolder-name}/(.*)$ /$1 break;
> ```
> 
> - where `{subfolder-name}` should be replaced with your desired subfolder name.
> 
> In my case, a complete Nginx `location` block would look like this:
> 
> ```nginx
> location /directus {
>     rewrite ^/directus/(.*)$ /$1 break;
>     proxy_pass http://localhost:3001;
>     proxy_http_version 1.1;
>     proxy_set_header Upgrade $http_upgrade;
>     proxy_set_header Connection 'upgrade';
>     proxy_set_header Host $host;
>     proxy_cache_bypass $http_upgrade;
> }
> ```
> 
> The `location` block is unaffected by the `PUBLIC_URL` configuration as this basically just refers to URL building in mails and other route building functions.
> 
> I hope the answer above helps others who are trying to achieve similar setup.

](https://github.com/directus/directus/discussions/4480#discussioncomment-473036)

Thanks @Repox. It worked for accessing the API.
However the admin dashboard wasn't loading due to static assets giving 404.
Fixed it by adding this block.
```
location /admin {
    proxy_pass http://0.0.0.0:8055;
}
```

Directus version `10.3.0`
